src/platform/cc13x2_26x2/crypto/sha256_alt.c

```diff
@@ -0,0 +1,193 @@
+ *  ti_drivers.h structure.
+ */
+const SHA2CC26X2_HWAttrs sha2CC26X2HWAttrs_sha = {
+    .intPriority = ~0,
+};
+
+void mbedtls_sha256_init(mbedtls_sha256_context * ctx)
+{
+    SHA2_Params sha2Params;
+
+    memset(ctx, 0, sizeof(mbedtls_sha256_context));
+    SHA2_Params_init(&sha2Params);
+
+    sha2Params.returnBehavior = SHA2_RETURN_BEHAVIOR_POLLING;
+
+    ctx->config.object  = &ctx->object;
+    ctx->config.hwAttrs = &sha2CC26X2HWAttrs_sha;
+
+    ctx->hndl = SHA2_construct(&ctx->config, &sha2Params);
+}
+
+void mbedtls_sha256_free(mbedtls_sha256_context * ctx)
+{
+    if (NULL == ctx)
+    {
+        return;
+    }
+
+    if (ctx->hndl != NULL)
+    {
+        SHA2_close(ctx->hndl);
+
+        memset(ctx, 0, sizeof(mbedtls_sha256_context));
+    }
+}
+
+/*
+ * Deviation from documentation in sha256.h.
+ *
+ * It has been noticed that some code using mbedtls does not call `init` before
+ * starting a SHA operation. Rather these implementations assume the `ctx`
+ * structure may be zero'ed out as initialization. To support this usage the
+ * driver instance is assumed to be uninitialized if the context's `hndl` is a
+ * NULL pointer. Start will initialize the driver and assign the handle to the
+ * context in this case.
+ */
+int mbedtls_sha256_starts_ret(mbedtls_sha256_context * ctx, int is224)
+{
+    SHA2_HashType type;
+
+    // open the driver here because no-one called init
+    if (ctx->hndl == NULL)
+    {
+        SHA2_Params sha2Params;
+
+        SHA2_Params_init(&sha2Params);
+
+        sha2Params.returnBehavior = SHA2_RETURN_BEHAVIOR_POLLING;
+
+        ctx->config.object  = &ctx->object;
+        ctx->config.hwAttrs = &sha2CC26X2HWAttrs_sha;
+
+        ctx->hndl = SHA2_construct(&ctx->config, &sha2Params);
+
+        if (NULL == ctx->hndl)
+        {
+            return MBEDTLS_ERR_SHA256_HW_ACCEL_FAILED;
+        }
+    }
+    else
+    {
+        SHA2_reset(ctx->hndl);
+    }
+
+    if (is224)
+    {
+        type = SHA2_HASH_TYPE_224;
+    }
+    else
+    {
+        type = SHA2_HASH_TYPE_256;
+    }
+
+    SHA2_setHashType(ctx->hndl, type);
+
+    return 0;
+}
+
+/**
+ * Deviation from documentation in sha256.h.
+ *
+ * It has been noticed that some code using mbedtls does not call `init` on the
+ * contexts being passed into this function. This seems to be done to avoid
+ * allocating hardware to cloned context instances. To support this use-case,
+ * only internal buffers to the SHA2 structure are copied without checking the
+ * driver handles.
+ */
+void mbedtls_sha256_clone(mbedtls_sha256_context * dst, const mbedtls_sha256_context * src)
+{
+    dst->object.hashType       = src->object.hashType;
+    dst->object.bytesInBuffer  = src->object.bytesInBuffer;
+    dst->object.bytesProcessed = src->object.bytesProcessed;
+    memcpy(dst->object.digest, src->object.digest, sizeof(dst->object.digest));
+    memcpy(dst->object.hmacDigest, src->object.hmacDigest, sizeof(dst->object.hmacDigest));
```

Comment:
hmac != SHA256. Why carry HMAC context when SHA256 is used? The HMAC context is what should carry its digest context.

Reply:
Those two buffers are landing zones for the driver's final digest. You're right I don't think I need to copy them when cloning an intermediate context.